sycl/include/sycl/ext/oneapi/experimental/group_load_store.hpp

```diff
@@ -106,6 +108,106 @@ int get_mem_idx(GroupTy g, int vec_or_array_idx) {
     return g.get_local_linear_id() +
            g.get_local_linear_range() * vec_or_array_idx;
 }
+
+// SPIR-V extension:
+// https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_subgroups.asciidoc,
+// however it doesn't describe limitations/requirements. Those seem to be
+// listed in the Intel OpenCL extensions for sub-groups:
+// https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups.html
+// https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_char.html
+// https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_long.html
+// https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups_short.html
+// Reads require 4-byte alignment, writes 16-byte alignment. Supported
+// sizes:
+//
+// +--------+-------------+
+// | uchar  | 1,2,4,8,16  |
+// | ushort | 1,2,4,8     |
+// | uint   | 1,2,4,8     |
+// | ulong  | 1,2,4,8     |
+// +--------+-------------+
+//
+// Utility type traits below are used to map user type to one of the block
+// read/write types above.
+
+template <typename value_type, std::size_t ElementsPerWorkItem, bool blocked>
+struct BlockInfo {
+  static constexpr int block_size =
+      sizeof(value_type) * (blocked ? ElementsPerWorkItem : 1);
+  static constexpr int num_blocks = blocked ? 1 : ElementsPerWorkItem;
+  static constexpr bool has_builtin =
+      detail::is_power_of_two(block_size) &&
+      detail::is_power_of_two(num_blocks) && block_size <= 8 &&
+      (num_blocks <= 8 || (num_blocks == 16 && block_size == 1));
+};
+
+template <typename BlockInfoTy> struct BlockTypeInfo;
+
+template <typename value_type, std::size_t ElementsPerWorkItem, bool blocked>
+struct BlockTypeInfo<BlockInfo<value_type, ElementsPerWorkItem, blocked>> {
+  using BlockInfoTy = BlockInfo<value_type, ElementsPerWorkItem, blocked>;
+  static_assert(BlockInfoTy::has_builtin);
+
+  using block_type = detail::cl_unsigned<BlockInfoTy::block_size>;
+  using block_pointer_type = const typename detail::DecoratedType<
```

Comment:
This `const` shouldn't be unconditional. Surprisingly to me `std::iterator_traits<const int *>::value_type` is `int` and not `const int`, so I'm looking into how to make that happen.